pages/docs/v3/core-apis/data-types.md

```diff
@@ -0,0 +1,116 @@
+
+As an example, consider the following object being passed to a Capacitor plugin call:
+
+```typescript
+{ 'foo': null, 'bar': [1, 2, null, 4]}
+```
+
+#### Dictionaries
+
+`CAPPluginCall` stores this data as its `options` dictionary property but has a variety of convenience accessors that operate on it. The convenience accessors will try to cast the value to the expected type(s) so `NSNull` values will get filtered out.
+
+```swift
+if let value = call.getString("foo") {
+    // false, `value` is nil
+}
+```
+
+However, accessing the storage property directly can return an `NSNull` object so its type will need to be checked before it is used.
+
+```swift
+let value = call.options["foo"]
+if let value = value {
+    // true, 'value' is a valid object
+}
+if let value = value as? Int {
+    // false, the NSNull object cannot be cast to Int
+}
+if let value = value as? NSNull {
+    // true, 'value' is a NSNull
+}
+```
+
+In general, it is not recommended to rely on the presence of a key to convey meaning. Only the corresponding value should be evaluated after its expected type is confirmed. Since dictionaries typically require the typing of each value as it is extracted, the inclusion of `NSNull` can mostly be ignored.
+
+#### Arrays
+
+Since accessing arrays typically requiring typing the whole collection, it is more important to consider if they are heterogeneous or not.
+
+```swift
+if let values = call.getArray("bar") {
+    // true, the array is all valid objects
+}
+if let values = call.getArray("bar", Int) {
+    // false, the array is a mix of Int and NSNull objects and can't be cast to [Int]
+}
+if let values = call.getArray("bar", Int?) {
+    // false, the array is a mix of Int and NSNull objects and can't be cast to [Int?]
+}
+```
+
+To help with this behavior, Capacitor includes a convenience extension that can map a heterogeneous array with `NSNull` values into an array of optionals. It works on the `JSValue` protocol, which represents all of the valid types that can be bridged between environments, but can be cast to a specific subtype.
+
+```swift
+if let values = call.getArray("bar").capacitor.replacingNullValues() as? [Int?] {
+    // true, 'values' is now cast to [Int?] with 'nil' at index 2
+}
+```
+
+---
+
+### Dates
+
+Data moving from the web runtime to native iOS code uses a different mechanism than data going in the other direction. In practice, this difference doesn't matter except for one data type: dates. The WebView can automatically translate JavaScript `Date` objects into native `NSDate` or `Date` objects but dates being returned from a plugin must be serialized as a string.
```

Comment:
This peculiarity makes me think we should consider switching to JSON serialization/deserialization for iOS. I think plugin developers are expecting it, and any automatic conversion of dates may be unexpected or at least not often used. It would bring parity between the messages and data types between iOS and Android (and other platforms) and I don't think we'll suffer that much of a performance hit (but would be worth benchmarking).

Reply:
The problem is that we cannot control how the webview passes the data and it's already handling the conversion. So we could:
1. Reject the payload if it contains dates. But it's not clear how best to message that back to the web view.
2. Serialize/deserialize the dates which doesn't really change anything.
3. Serialize the dates into strings before passing them to the plugin but then we're just doing a different magic conversion, one that might introduce side effects.

If anything, what we'd probably want to do is automatically handle serializing dates into strings when passing data back to the web but it would not be trivial to handle deserializing those back into JavaScript Dates. So we'd still have a mismatch where data transit is different in different directions.

Reply:
> we cannot control how the webview passes the data 

Sure we can--all Capacitor calls go through `postToNative`: https://github.com/ionic-team/capacitor/blob/main/core/src/bridge.ts#L37-L44

For Android we run `JSON.stringify()`: https://github.com/ionic-team/capacitor/blob/main/core/src/bridge.ts#L30

Reply:
@dwieeb After discussing this issue with @theproducer, another possibility would be to handle this natively. That is, we examine the call payload and turn any `NSDate` objects into ISO 8901 strings at the time that [we coerce the types](https://github.com/ionic-team/capacitor/blob/9ff9d1b07e6d6f44781e580183475af9bd8e2e9b/ios/Capacitor/Capacitor/CapacitorBridge.swift#L389).

Reasons
1. Consistency since dates will automatically get turned into strings on both platforms
2. The NSDate-object behavior can be preserved for any Cordova plugins that rely on it (since we've identified the plugin type at that point)
3. The performance impact should be less than stringifying the whole blob in JS first

We could also do the same when serializing data back to JS. So Dates would automatically get stringified in both directions rather than requiring that developers do it (although they'd be free to pre-stringify the date if they wanted some other format).

Reply:
Sounds good to me.